File: PS4/tourist.py
INF = 10**18

class min_cost_max_flow:
  def __init__(self, V):
    self.V = V
    self.EL = []
    self.AL = [list() for _ in range(V)]
    self.vis = [False] * V
    self.total_cost = 0
    self.d = None
    self.last = None

  def SPFA(self, s, t):
    self.d = [INF] * self.V
    self.d[s] = 0
    self.vis[s] = True
    q = [s]
    while len(q) != 0:
      u = q[0]
      q.pop(0)
      self.vis[u] = False
      for idx in self.AL[u]:
        v, cap, flow, cost = self.EL[idx]
        if cap-flow > 0 and self.d[v] > self.d[u]+cost:
          self.d[v] = self.d[u]+cost
          if not self.vis[v]:
            q.append(v)
            self.vis[v] = True
    return self.d[t] != INF

  def DFS(self, u, t, f=INF):
    if u == t or f == 0:
      return f
    self.vis[u] = True
    for i in range(self.last[u], len(self.AL[u])):
      v, cap, flow, cost = self.EL[self.AL[u][i]]
      if not self.vis[v] and self.d[v] == self.d[u]+cost:
        pushed = self.DFS(v, t, min(f, cap-flow))
        if pushed != 0:
          self.total_cost += pushed * cost
          flow += pushed
          self.EL[self.AL[u][i]][2] = flow
          rv, rcap, rflow, rcost = self.EL[self.AL[u][i]^1]
          rflow -= pushed
          self.EL[self.AL[u][i]^1][2] = rflow
          self.vis[u] = False
          self.last[u] = i
          return pushed
    self.vis[u] = False
    return 0

  def add_edge(self, u, v, w, c, directed=True):
    if u == v:
      return
    self.EL.append([v, w, 0, c])
    self.AL[u].append(len(self.EL)-1)
    self.EL.append([u, 0 if directed else w, 0, -c])
    self.AL[v].append(len(self.EL)-1)

  def mcmf(self, s, t):
    mf = 0
    while self.SPFA(s, t):
      self.last = [0] * self.V
      f = self.DFS(s, t)
      while f != 0:
        mf += f
        f = self.DFS(s, t)
    return mf, self.total_cost


def main():
  print("Reading cases...")
  cases = int(input().strip())
  print("Read cases.")
  for ii in range(0, cases):
    print("New test case")
    W, H = map(int, input().strip().split(" "))
    mcmf = min_cost_max_flow(W*H*2+2)
    s = W*H*2
    t = W*H*2+1
    mcmf.add_edge(s, 0, INF, 0, False)
    mcmf.add_edge(W*H*2-1, t, INF, 0, False)
        
    print("Reading input...")
    graph = [[] for ff in range(0, H)]
    for xx in range(0, H):
      for yy, cell in enumerate(input().strip()):
        graph[xx].append(cell)
        
    print("Building graph...")
    for xx in range(0, H):
      for yy in range(0, W):
        inNode = xx*W+yy
        outNode = inNode + W*H

        if(graph[xx][yy] == '*'):
          mcmf.add_edge(inNode, outNode, 1, -1)
        else:
          mcmf.add_edge(inNode, outNode, 1, 0)

        neighbours = [(xx+1, yy), (xx, yy+1)]
        for (uu, vv) in neighbours:
          if((uu>=0) and (uu<H) and (vv>=0) and (vv<W)) and (graph[uu][vv] != '#'):
            inIndex = uu*W+vv
            outIndex = inIndex + W*H
            mcmf.add_edge(outNode, inIndex, 1, 0)
            mcmf.add_edge(outIndex, inNode, 1, 0)
    print("Finished building graph.")
    print("Starting MCMF computation...")
    mf, cost = mcmf.mcmf(s, t)
    print("Finished MCMF computation.")
    print(-cost)
            
if __name__ == '__main__':
  main()

File: PS4/tourist.cpp
/*Courtesy of Steven Halim, Felix Halim, Suhendry Effendy
CP4 Free Source Code Project (cpp/gnu++17, java/java11, py/python3, and ml/ocaml).
https://github.com/stevenhalim/cpbook-code/blob/master/ch9/mcmf.py

-- Python implementation of library code translated to C++ --
*/

#include <bits/stdc++.h>

using namespace std;

const long long INF = 1e18;

struct Edge {
    int v;
    long long cap, flow, cost;
};

class MinCostMaxFlow {
    int V;
    vector<Edge> EL;
    vector<vector<int>> AL;
    vector<bool> vis;
    vector<long long> d;
    vector<int> last;
    long long total_cost = 0;

public:
    MinCostMaxFlow(int V) : V(V), AL(V), vis(V, false), d(V, INF) {}

    bool SPFA(int s, int t) {
        fill(d.begin(), d.end(), INF);
        d[s] = 0;
        vis[s] = true;
        queue<int> q;
        q.push(s);

        while (!q.empty()) {
            int u = q.front();
            q.pop();
            vis[u] = false;

            for (int idx : AL[u]) {
                int v = EL[idx].v;
                long long cap = EL[idx].cap, flow = EL[idx].flow, cost = EL[idx].cost;
                
                if (cap - flow > 0 && d[v] > d[u] + cost) {
                    d[v] = d[u] + cost;

                    if (!vis[v]) {
                        q.push(v);
                        vis[v] = true;
                    }
                }
            }
        }

        return d[t] != INF;
    }

    long long DFS(int u, int t, long long f = INF) {
        if (u == t || f == 0) return f;
        vis[u] = true;

        for (int &i = last[u]; i < AL[u].size(); ++i) {
            int idx = AL[u][i];
            Edge &e = EL[idx], &re = EL[idx^1];
            
            if (!vis[e.v] && d[e.v] == d[u] + e.cost && e.cap - e.flow > 0) {
                long long pushed = DFS(e.v, t, min(f, e.cap - e.flow));

                if (pushed != 0) {
                    total_cost += pushed * e.cost;
                    e.flow += pushed;
                    re.flow -= pushed;

                    vis[u] = false;
                    return pushed;
                }
            }
        }

        vis[u] = false;
        return 0;
    }

    void add_edge(int u, int v, long long w, long long c, bool directed = true) {
        if (u == v) return;

        EL.push_back({v, w, 0, c});
        AL[u].push_back(EL.size() - 1);
        
        EL.push_back({u, directed ? 0 : w, 0, -c});
        AL[v].push_back(EL.size() - 1);
    }

    pair<long long, long long> mcmf(int s, int t) {
        long long mf = 0;
        while (SPFA(s, t)) {
            last.assign(V, 0);
            long long f;

            while ((f = DFS(s, t)) != 0) {
                mf += f;
            }
        }

        return {mf, total_cost};
    }
};

int main() {
    int cases;
    cin >> cases;

    while (cases--) {
        int W, H;
        cin >> W >> H;
        int WH = W * H;
        
        MinCostMaxFlow mcmf(WH * 2 + 2);
        int s = WH * 2, t = WH * 2 + 1;
        
        mcmf.add_edge(s, 0, INF, 0, false);
        mcmf.add_edge(WH * 2 - 1, t, INF, 0, false);

        vector<string> graph(H);

        for (int i = 0; i < H; ++i) {
            cin >> graph[i];
        }

        for (int x = 0; x < H; ++x) {
            for (int y = 0; y < W; ++y) {
                int inNode = x * W + y;
                int outNode = inNode + WH;

                if (graph[x][y] == '*') {
                    mcmf.add_edge(inNode, outNode, 1, -1);
                    mcmf.add_edge(inNode, outNode, 1, 0);
                } else if (graph[x][y] != '#') {
                    mcmf.add_edge(inNode, outNode, 2, 0);
                }

                vector<pair<int, int>> neighboursSE = {{x + 1, y}, {x, y + 1}};

                for (auto [u, v] : neighboursSE) {
                    if (u < H && v < W && graph[u][v] != '#') {
                        int inIndex = u * W + v;
                        mcmf.add_edge(outNode, inIndex, 2, 0);
                    }
                }
            }
        }

        auto [mf1, cost1] = mcmf.mcmf(s, t);
        cout << -cost1 << endl;
    }

    return 0;
}